src/pushsource/_impl/backend/koji_source.py

```diff
@@ -283,18 +284,38 @@ def _push_items_from_rpm_meta(self, rpm, meta):
 
         candidate_paths = []
 
-        # If signing keys requested, try them in order of preference
-        for key in self._signing_key:
-            if key:
-                key = key.lower()
-                candidate = os.path.join(build_path, self._pathinfo.signed(meta, key))
-            else:
-                candidate = unsigned_path
-            candidate_paths.append(candidate)
-            if os.path.exists(candidate):
-                rpm_path = candidate
-                rpm_signing_key = key
+        timeout = int(os.getenv("PUSHSOURCE_SRC_POLL_TIMEOUT") or "0")
+        poll_rate = int(os.getenv("PUSHSOURCE_SRC_POLL_RATE") or "30")
+        max_attempts = timeout // poll_rate
+
+        for i in range(max_attempts + 1):
+            # We only want to poll if signing key is requested
+            if not self._signing_key:
                 break
+            # If signing keys requested, try them in order of preference
+            for key in self._signing_key:
+                if key:
+                    key = key.lower()
+                    candidate = os.path.join(
+                        build_path, self._pathinfo.signed(meta, key)
+                    )
+                else:
+                    candidate = unsigned_path
+                if candidate not in candidate_paths:
+                    candidate_paths.append(candidate)
+                if os.path.exists(candidate):
+                    rpm_path = candidate
+                    rpm_signing_key = key
+                    break
+
+            if rpm_path or i == max_attempts:
+                break
+            LOG.info(
+                "Waiting for %s seconds for any of the following paths to appear: %s",
+                poll_rate,
+                ", ".join(candidate_paths),
+            )
+            time.sleep(poll_rate)
```

Comment:
I'm not sure the implemented logic is safe unfortunately.  Depending on exactly how the delayed propagation in content works.

Consider for example if two signing keys A and B are requested, in order.  Imagine that in this case, RPMs are available for both keys, but there's a delay in the content becoming available.

Isn't it possible that you test for existence of A and get False, and then test for existence of B and get True, just because B happened to sync before A or even just because some time passed between your checks on A and B?  So you end up picking B even though you should pick A.  And this is going to happen at random, meaning the overall behavior is going to be slightly flaky.

It sucks, but if we want this to be reliable I think we need to wait the entire timeout while looking only for the *first* file.  Otherwise if you find any file other than the first, you still don't know if the first file is genuinely missing or it's just about to appear.  Basically if we are saying that PUSHSOURCE_SRC_POLL_TIMEOUT is the max possible synchronization delay we expect between storage, then it is not possible to safely conclude that any particular file doesn't exist until after that timeout has been reached.

Alternatively: does Pub actually need to use this with more than one signing_key?  If not, then it would be an option to implement the polling only in the case with a single signing_key so there's only one path to check. Or even to add an option to this source which makes it return the paths without any existence check at all, which also fixes the problem that this polling is going to happen more than once in the case where a file is missing.

Reply:
Yeah, this did occur to me during implementation, but it seemed like the code didn't care that much about which key will be used in the end.

I changed the logic to poll only for the first key until timeout. Though if the highest prio key fails to appear, a wrong key will be used (instead of push failing, which will happen in other cases of polling).